### ISSUE TITLE
Node#remove

### ISSUE THREAD

**Issue by ai** [postcss/postcss#191]
Now we has two qay to remove nodes: `parent.remove(node)` and `node.removeSelf()`. Maybe we should join them to just 'node.remove()' as in DOM API.

**Comment by MoOx** [comment]
parent.removeChild(node) or detachChild(), and node.remove() or node.detach().
Maybe detach is more clear cause node is not deleted in memory.

**Comment by jonathantneal** [comment]
I would use `remove` and mirror the DOM API, as you have nearly everywhere else.

**Comment by ben-eb** [comment]
+1 for `node.remove`. :+1:

**Comment by 1j01** [comment]
@MoOx "remove" doesn't strongly imply it's deleted in memory either, so it's fine.

**Comment by MoOx** [comment]
totally agree

**Comment by ai** [comment]
So `decl.remove()` and `rule.removeChild(decl)`?

I think 5.0 is a good time to deprecate `rule.remove()` to clean method for 6.0 refactoing.

**Comment by ben-eb** [comment]
@ai I thought we were talking about removing `removeSelf`. So a `remove()` with no params is removing that node, but a `remove(node)` removes the child node passed as the parameter.

**Comment by lydell** [comment]
@ben-eb Just curious, is `remove(node)` also inspired of the DOM method, or is that something you've come up with? Also, what is the benefit of running `parent.remove(node)` over simply `node.remove()`?

**Comment by ben-eb** [comment]
I should clarify; that `remove(node)` is really `parent.remove(node)`, i.e. the existing API. It's still worth using that method if you are iterating `rules` instead of `decls` for example.

**Comment by lydell** [comment]
But even then you need a reference to `node` so you could just as well have called `node.remove()`?

**Comment by ben-eb** [comment]
Yeah, you're right. In this case why would we need `removeChild(decl)`?

**Comment by ai** [comment]
`node.parent.removeChild(index)` is faster than `node.remove()` or `node.parent.removeChild(node)`.

**Comment by lydell** [comment]
Ah, I had no idea you could remove nodes by index.

What about having `node.remove()` (never takes any arguments) for removing by node reference, and `parent.removeChildByIndex(index)` for fast removals when you've got the index available?

**Comment by ai** [comment]
I am OK to split `removeChild` method, but `removeChildByIndex` is too long, looks like good old Java :D.

**Comment by ai** [comment]
BTw, I think `removeChildByIndex` is low-level optimization. Most of developers can miss it.

**Comment by ai** [comment]
Done by @ben-eb
